lib/puppet/network/http/factory.rb

```diff
@@ -38,6 +38,11 @@ def create_connection(site)
     http.read_timeout = Puppet[:http_read_timeout]
     http.open_timeout = Puppet[:http_connect_timeout]
 
+    if Puppet[:sourceaddress] && http.respond_to?(:local_host)
+      Puppet.debug("Using source IP #{Puppet[:sourceaddress]}")
+      http.local_host = Puppet[:sourceaddress]
+    end
```

Comment:
After thinking some more about this, I think we want to raise if it doesn't respond to the method, otherwise, we might wonder why it's not working as intended.

Reply:
Added an exception and testcase.